Astro (SSR + DB) + Htmx + Netlify Blogpost (#432)
hendriknielaender/double-trouble: data/blog/howto-htmx-astrodb-astrossr.md

---
publishDate: "Apr 2 2024"

> title: "How to ditch React and build an app with HTMX, Astro DB & Astro SSR"

Reviewer: ```suggestion
title: "Ditch React: Build Faster with htmx, Astro DB & Astro SSR in 2024"
```

Reviewer: Ensure htmx is always lowercase. https://htmx.org/

description: "Explore common techniques and solutions needed for building a basic CRUD app with this modern react-less stack."
image: "~/assets/images/thumbnails/htmx_app.jpg"
imageCreditUrl: https://midjourney.com
tags: [htmx, react, astro, astrodb, astrossr, crud, netlify]
---

These days, many people feel tired of React, and as [Primeagen](https://twitter.com/ThePrimeagen)
announced it: `This is the year of HTMX!` While the htmx examples in the documentation look easy

> enough, we took it to the test and and re-implement an existing React project in HTMX.

Reviewer: ```suggestion
enough, we took it to the test and re-implemented an existing React project in htmx.
```

We had created the original app earlier this year mostly to play around with Websockets and
AppSync subcriptions. In the frontend it was based on React and Nextjs, hosted ontop of Vercel. On
the backend it ran on AWS AppSync and Dynamodb as a persistent storage. In the re-implementation

> we moved the frontend from to purely html, using [HTMX](https://htmx.org/) and

Reviewer: ```suggestion
we moved the frontend from to purely html, using [htmx](https://htmx.org/ "Visit HTMX Official Site for more details") and
```

[Astro](https://astro.build). For the backend we used [Astro DB](https://astro.build/db/), which
is currently in early preview, and Astro SSR hosted on [Netlify](https://www.netlify.com/) in the
free tier.

The basic point of this experiment was to see, how CRUD apps could be implemented in a
straightforward way, but also to see how the experience would be with HTMX instead of React.

## Responsive Buttons

A responsive button would change its appearance while the operation is ongoing. For our usecase
this means we want to show a "Confirm" text while the button is ready, and a loading spinner while
the confirmation is being processed.

<img
    style="display: block;
           margin-left: auto;
           margin-right: auto;"

>     src="../../src/assets/images/posts/howto-htmx-astrodb-astrossr/ezgif-button.gif">

Reviewer: ```suggestion
  alt="Demonstration of HTMX responsive button"
src="../../src/assets/images/posts/howto-htmx-astrodb-astrossr/ezgif-button.gif">
```

---
publishDate: "May 23 2024"
title: "Ditch React: Build Faster with htmx, Astro DB & Astro SSR in 2024"

> description: "Explore common techniques and solutions needed for building a basic CRUD app with this modern react-less stack."

Reviewer: ```suggestion
description: "Explore common techniques and solutions for building a fast and efficient CRUD app using htmx, Astro DB, and Astro SSR without React in 2024."
```

image: "~/assets/images/thumbnails/htmx_app.jpg"
imageCreditUrl: https://midjourney.com
tags: [htmx, react, astro, astrodb, astrossr, crud, netlify]
---

These days, many people feel tired of React, and as [@ThePrimeagen](https://twitter.com/ThePrimeagen)
announced it:

> This is the year of HTMX!

While the htmx examples in the [documentation](https://htmx.org/examples/) look easy enough, we
decided to take it to the test and re-implemented an existing [React
project](https://github.com/flyck/party-task-planner) in
[htmx](https://github.com/flyck/astro-party).

We had created the original app earlier this year, mostly to play around with Websockets and

> AppSync subscriptions. In the frontend it was based on React and Next.js, hosted on top of

Reviewer: ```
In the frontend, it was based on React and Next.js, hosted on Vercel. The backend ran on AWS AppSync and DynamoDB for persistent storage. In the re-implementation, we moved the frontend to pure HTML and JavaScript using htmx and Astro. For the backend, we used Astro DB and Astro SSR, hosted on Netlify.
```
Rewrite suggestion.

Vercel. On the backend it ran on AWS AppSync and Dynamodb as a persistent storage. In the
[re-implementation](https://github.com/flyck/astro-party), we moved the frontend to pure HTML and
js, using [htmx](https://htmx.org/ "Visit HTMX Official Site for more details") and
[Astro](https://astro.build). For the backend we used [Astro DB](https://astro.build/db/), which
is currently in early preview, and Astro SSR hosted on [Netlify](https://www.netlify.com/).

> We were curious how well this stack could could work for CRUD apps, but especially the experience

Reviewer: ```suggestion
We were curious how well this stack could work for CRUD apps, but especially the experience
```

of using htmx instead of React. In the following we'll share basic techniques for this stack,
and also our conclusions, looking back.

> ## Responsive Buttons

Reviewer: ```suggestion
## Table of Contents
- [Creating Responsive Buttons with htmx and Tailwind CSS](#creating-responsive-buttons-with-htmx-and-tailwind-css)
- [Implementing Skeletons for Better UX](#implementing-skeletons-for-better-ux)
- [Inline Validation for Immediate Feedback](#inline-validation-for-immediate-feedback)
- [Handling Errors with Toast Notifications](#handling-errors-with-toast-notifications)
- [Enhancing Navigation with View Transitions](#enhancing-navigation-with-view-transitions)
- [Using Astro SSR & Netlify for Hosting](#using-astro-ssr--netlify-for-hosting)
- [Is htmx the Ultimate Choice?](#is-htmx-the-ultimate-choice)
- [Conclusion](#conclusion)

## Creating Responsive Buttons with htmx and Tailwind CSS
```

A responsive button would change its appearance while the operation is ongoing. For our use-case
this means we want to show a "Confirm" text while the button is ready, and a loading spinner while
the confirmation is being processed.

<img
    style="display: block;
           margin-left: auto;
           margin-right: auto;"
  alt="Demonstration of HTMX responsive button"
src="../../src/assets/images/posts/howto-htmx-astrodb-astrossr/button.gif">
</img>

> Htmx automatically adds predefined CSS classes to requesting elements, so with tailwinds

Reviewer: ```suggestion 
Htmx automatically adds predefined CSS classes to requesting elements, so with tailwind's
```

`.astro` syntax for templating all around. The experience here for our small app was simply
amazing. The complexity of React and Next.js could be completely skipped, and we could deliver an
app without any JSON contracts minimal dependencies.

## Is htmx the ultimate choice?

Htmx author Carson himself points out the advantages and disadvantages of htmx based on the
situation (see [essay](https://htmx.org/essays/when-to-use-hypermedia/)).

To us, htmx excels at going back to basics. HTML forms are a breeze, and the whole contracted
JSON api middleman can be skipped, by shipping HTML directly. Especially with Astro, having
`.astro` templates all the way through, with minimal external dependencies, keeps the complexity
low.

On the other hand, building this app highlighted also a few strengths of React. Handling the data
flow between components is the main purpose of the React component properties. With htmx, doing
the same while only fetching data once, had us come up with some last-mile JavaScript. The
alternative could be to simply fetch more HTML, which seems to be part of the htmx paradigm.

> Generally the chance of writing explicit JavaScript seems slightly higher with htmx, but at-least

Reviewer: ```suggestion
Generally the chance of writing explicit JavaScript seems slightly higher with htmx, but at least
```

you know what you're doing, and it's very honest work. Though in this case more explicit JavaScript
was written, the perceived complexity was lower and the overall enjoyment greater.

## Rounding up

> Coming from react, htmx introduces a completely different application paradigm. It requires going

Reviewer: ```suggestion
Coming from React, htmx introduces a completely different application paradigm. It requires going
```

back to the basics, which felt good. Simply not having the app written in React felt great.

Also the combination of [Astro DB](https://astro.build/db/) with [Astro
SSR](https://docs.astro.build/en/guides/server-side-rendering/) hosted on
[Netlify](https://www.netlify.com/) was great. After it initially seemed like an anti-pattern to
use Astro for dynamic content, the site itself is quite fast and the experience was great.

> Htmx is conceptionally easy to grasp, and while the initial learning curve is somewhat steep, it

Reviewer: ```suggestion
Htmx is conceptually easy to grasp, and while the initial learning curve is somewhat steep, it
```
